hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java

```diff
@@ -537,7 +558,9 @@ private void addPendingEvent(Event event) {
         public void run() {
             int numTimeouts = checkPendingTimeouts();
 
-            if (!pendingData.isEmpty() || !pendingEvents.isEmpty() || numTimeouts > 0) {
+            checkMissingStates();
```

Comment:
This isn't quite what I was suggesting.  Let me try to clarify, the idea is closer to what we do with `checkPendingTimeouts()`.   `MissingStates` is a `Set` of `MissingState` with one entry per `MissingCondition` and an associated Fact in working memory.   This should not be a global shared with the rules engine.  Instead, we'd want a new set, like `Set<MissingState> updatedMissingState<>()`, that would be the shared global.

So, here, instead of:
`checkMissingStates();`
We'd have something like:
`int numMissingTooLong = checkMissingStates();`

The `checkMissingStates()` method would do a couple of things:

1. It would run through (the new global) `updatedMissingStates` and reflect those update times in the relevant `MissingStates` entry.
2. It would clear `updatedMissingStates` in preparation of the next engine run.
3. It would run through `MissingStates` and for any that had surpassed the defined interval of inactivity it would generate the positive `MissingConditionEval` fact and inject it into working memory.  And, in this case probably also reset the interval timer.

So, the check:
`if (!pendingData.isEmpty() || !pendingEvents.isEmpty() || numTimeouts > 0 || !missingStates.isEmpty())`
would change to:
`if (!pendingData.isEmpty() || !pendingEvents.isEmpty() || numTimeouts > 0 || numMissingTooLong > 0`